sacred/observers/s3_observer.py

```diff
@@ -0,0 +1,328 @@
+
+from sacred.commandline_options import CommandLineOption
+from sacred.dependencies import get_digest
+from sacred.observers.base import RunObserver
+from sacred.serializer import flatten
+import re
+import socket
+
+DEFAULT_S3_PRIORITY = 20
+
+
+def _is_valid_bucket(bucket_name):
+    # See https://docs.aws.amazon.com/awscloudtrail/latest/userguide/
+    # cloudtrail-s3-bucket-naming-requirements.html
+    if len(bucket_name) < 3 or len(bucket_name) > 63:
+        return False
+    if '..' in bucket_name or '.-' in bucket_name or '-.' in bucket_name:
+        return False
+    for char in bucket_name:
+        if char.isdigit():
+            continue
+        if char.islower():
+            continue
+        if char == '-':
+            continue
+        return False
+    try:
+        # If a name is a valid IP address, it cannot be a bucket name
+        socket.inet_aton(bucket_name)
+    except socket.error:
+        return True
+
+
+class S3FileObserver(RunObserver):
+    VERSION = 'S3FileObserver-0.1.0'
+
+    @classmethod
+    def create(cls, bucket, basedir, resource_dir=None, source_dir=None,
+               priority=DEFAULT_S3_PRIORITY):
+        resource_dir = resource_dir or os.path.join(basedir, '_resources')
+        source_dir = source_dir or os.path.join(basedir, '_sources')
+
+        return cls(bucket, basedir, resource_dir, source_dir, priority)
+
+    def __init__(self, bucket, basedir, resource_dir, source_dir,
+                 priority=DEFAULT_S3_PRIORITY):
+        if not _is_valid_bucket(bucket):
+            raise ValueError("Your chosen bucket name does not follow AWS "
+                             "bucket naming rules")
+
+        self.basedir = basedir
+        self.bucket = bucket
+        # Keeping the convention of referring to locations in S3 as `dir`
+        # because that is a useful mental model and there isn't a better word
+        self.resource_dir = resource_dir
+        self.source_dir = source_dir
+        self.priority = priority
+        self.dir = None
+        self.run_entry = None
+        self.config = None
+        self.info = None
+        self.cout = ""
+        self.cout_write_cursor = 0
+        self.s3 = boto3.resource('s3')
+        self.saved_metrics = {}
+
+    def _objects_exist_in_dir(self, prefix):
+        # This should be run after you've confirmed the bucket
+        # exists, and will error out if it does not exist
+
+        bucket = self.s3.Bucket(self.bucket)
+        all_keys = [el.key for el in bucket.objects.filter(Prefix=prefix)]
+        return len(all_keys) > 0
+
+    def _list_s3_subdirs(self, prefix=None):
+        if prefix is None:
+            prefix = self.basedir
+        try:
+            bucket = self.s3.Bucket(self.bucket)
+            all_keys = [el.key for el in bucket.objects.filter(Prefix=prefix)]
+        except ClientError as er:
+            if er.response['Error']['Code'] == 'NoSuchBucket':
+                return None
```

Comment:
I would not handle this error in this way. Usually I would say return an empty list, if you cannot find anything. However, in this case we want to differentiate between an empty bucket and a non-existing bucket. So I would not handle this error in this method and instead make another method `bucket_exists(str) -> bool` to ask this question explicitly.

Reply:
Makes sense, have now made a separate bucket_exists method, which I agree is cleaner